shell/packages/sandstorm-ui-powerbox/powerbox-client.js

```diff
@@ -0,0 +1,479 @@
+SandstormPowerboxRequest = class SandstormPowerboxRequest {
+  constructor(db, requestInfo) {
+    check(requestInfo, Match.ObjectIncluding({
+      source: Match.Any,
+      origin: Match.Any,
+      rpcId: Match.Any,
+      // For reasons I don't understand, Match.Optional does not work here.
+      query: Match.OneOf(undefined, [Object]),
+      saveLabel: Match.Optional(Match.Any),
+      grainId: String,
+      identityId: String,
+      onCompleted: Function,
+    }));
+
+    this._db = db;
+    this._requestInfo = requestInfo;
+    this._completed = false;
+    this._error = new ReactiveVar(undefined);
+
+    // State to track UI status.
+    this._filter = new ReactiveVar("");
+    this._selectedProvider = new ReactiveVar(undefined);
+  }
+
+  finalize() {
+    if (!this._completed) {
+      // postMessage back to the origin frame that the request was cancelled.
+      this._requestInfo.source.postMessage({
+        rpcId: this._requestInfo.rpcId,
+        error: this._error.get() || "User cancelled request",
+      }, this._requestInfo.origin);
+    }
+  }
+
+  hex64ToDecimal(n) {
+    if (typeof n != "string") {
+        throw new Error("Expected string.");
+    }
+    if (n.length != 18 || n.lastIndexOf("0x", 0) !== 0) {
```

Comment:
Heh, now I'm reviewing my own code...

Should also validate that all characters are hexadecimal digits. And should verify that the `parseInt()` calls succeed.